torch/csrc/autograd/engine.cpp

```diff
@@ -1117,13 +1117,38 @@ void Engine::add_thread_pool_task(const std::weak_ptr<GraphTask>& graph_task) {
 }
 
 void GraphTask::init_to_execute(Node& graph_root, const edge_list& outputs, bool accumulate_grad) {
+  // Populates exec_info so nodes that should be executed have `exec_info[node].needed_ = true`
+  // Only nodes that have a path to any edge in `outputs` should be executed.
+  // The code below populates exec_info using recursion, but the actual code does this
+  // iteratively. Refer to the numbering to see how the actual code corresponds.
```

Comment:
I would add one more sentence here explaining that in the iterative version, when you are working with the current Node, you are reponsible to update your parent's is_needed after all your children have been updated.